benchmarks/fuzzy_join_benchmark.py

```diff
@@ -0,0 +1,299 @@
+
+    metric = distance_metrics()[similarity]
+    sim = metric(left_enc, right_enc)
+
+    idx_closest = np.argmin(sim, axis=1)
+    dist = sim[range(len(sim)), idx_closest]
+    dist_ord = np.argsort(np.ravel(dist))
+    for idx in left.index:
+        joined.loc[idx, "col_to_embed"] = right_clean[idx_closest[idx]]
+    for k in range(1, len(left), 10):
+        mask = dist_ord[:k]
+        joined_k = joined.iloc[mask]
+        pr, re, f1 = evaluate(
+            list(zip(joined_k["col_to_embed"], joined_k[right_col])),
+            list(zip(gt["title_l"], gt["title_r"])),
+        )
+        pr_list.append(pr)
+        re_list.append(re)
+        f1_list.append(f1)
+    return pr_list, re_list, f1_list
+
+
+def thefuzz_precision_recall(
+    left, right, gt, left_col, right_col, scorer=partial_ratio
+):
+    pr_list = []
+    re_list = []
+    f1_list = []
+    for threshold in range(60, 99, 15):
+        joined = thefuzz_merge(
+            left,
+            right,
+            left_col,
+            right_col,
+            threshold=threshold,
+            limit=1,
+            scorer=scorer,
+        )
+        pr, re, f1 = evaluate(
+            list(zip(joined[left_col], joined["matches"])),
+            list(zip(gt["title_l"], gt["title_r"])),
+        )
+
+        pr_list.append(pr)
+        re_list.append(re)
+        f1_list.append(f1)
+    return pr_list, re_list, f1_list
+
+
+def autofj_precision_recall(left, right, gt, n_points=20):
+    pr_list = []
+    re_list = []
+    f1_list = []
+    for target in np.linspace(0.6, 1, n_points):
+        autofj = AutoFJ(precision_target=target, verbose=True)
+        LR_joins = autofj.join(left, right, id_column="id")
+        if len(LR_joins) != 0:
+            gt_joins = gt[["id_l", "id_r"]].values
+            LR_joins = LR_joins[["id_l", "id_r"]].values
+            p, r, f1 = evaluate(LR_joins, gt_joins)
+            pr_list.append(p)
+            re_list.append(r)
+            f1_list.append(f1)
+            # print("Precision:", p, "Recall:", r, "F1:", f1)
+    return pr_list, re_list, f1_list
+
+
+def best_precision_recall(pr_list, re_list, n_bins=13):
+    bins = np.digitize(re_list, np.quantile(re_list, np.linspace(0, 1, n_bins)))
+    res_pr = np.zeros(n_bins - 1)
+    res_re = np.zeros(n_bins - 1)
+    for i in range(1, n_bins):
+        mask = bins == i
+        # print("Recall:", re_list[mask], "Precision:", pr_list[mask])
+        if len(pr_list[mask]) > 0:
+            res_pr[i - 1] = np.nanmax(pr_list[mask])
+            res_re[i - 1] = np.nanmean(re_list[mask])
+        else:
+            res_pr[i - 1] = np.nan
+            res_re[i - 1] = np.nan
+    return res_pr, res_re
+
+
+if __name__ == "__main__":
+    left_1, right_1, gt_1 = fetch_data("Country")
+
+    pr_list = []
+    re_list = []
+    for analyser in ["char", "char_wb"]:
+        for max_n_gram in [4]:  # [2, 3, 4]:
+            for similarity in ["cosine", "l1", "l2"]:
```

Comment:
Similarly: Would `itertools.product` be useful here, flattening loops and removing indentation levels?